### PR TITLE
ASoC: SOF: Intel: mtl/lnl: correct ROM status code checking and fix dsp dump

### PR DESCRIPTION
**Pull request by ujfalusi**
Hi,

the rom_status_reg was incorrect for MTL/LNL and because of that the target_state checking was not implemented which makes it impossible to handle boot failures and to apply fallback in case of IMR failure.

### CODE REVIEW

**Review by RanderWang — APPROVED** [review]
LGTM. Thanks !

**Comment by ujfalusi** [comment]
Looks like this is not working on LNL:
```
FW Poll Status: reg[0x160100]=0xffffffff timedout
```

**Comment by ujfalusi** [comment]
Changes since v1:
- revert the .rom_status_reg value changes, local testing passes with using the SRAM window 'registers'

**Comment by ujfalusi** [comment]
The revert do not work on LNL either:
```
 FW Poll Status: reg[0x180000]=0xffffffff timedout
```

**Comment by ujfalusi** [comment]
Using the SRAM window registers were complete disaster on MTL in PR test.

**Comment by ujfalusi** [comment]
Changes since v2:
- Use the same register on LNL as used on MTL (HFFLGP1QW0)
- Wrongly update the commit title for the squashed rom_status_reg update patch...

**Comment by ujfalusi** [comment]
v3 also fails on LNL:
```
FW Poll Status: reg[0x163200]=0x0 timedout
```
at least it is not reading 0xffffffff anymore. I still think that the original 0x160100 address should be the correct one. I wonder if we have some access restriction not lifted?

I'll wait for the CI to test MTL again, just in case.

**Comment by ujfalusi** [comment]
Changes since v4:
- Only wait for the target_state when the chip->rom_status_reg != MTL_DSP_ROM_STS, iow, only MTL for now.
- Add patch which should fix the recovery after failed IMR booting due to enabled interrupts intercepting the purge IPC done message.

**Comment by ujfalusi** [comment]
Changes since v5:
- LNL ROM status handling is fixed (host needs to read HFDSC/HFDEC registers)
- Fixes tags added to patches [1]
- Extended the PR with patches to correct the dsp dump in case of a panic on ACE architecture

[1] The fixes tags are pointing to the original patches, this might cause conflicts if additional patches are not present on top of them. This is true for the following two patches:
ASoC: SOF: Intel: mtl: Disable interrupts when firmware boot failed
ASoC: SOF: Intel: mtl: Implement firmware boot state check
There has been unrelated code changes around the places these patches touch but the original code was not correct.

I think the code needs to be fixed in any case, or should I use Stable: 6.6, 6.7 tag? Not sure how it backports to these.

**Comment by ujfalusi** [comment]
Verified to work on MTL family and LNL,

**Review by kv2019i — APPROVED** [review]

**Review by plbossart — APPROVED** [review]
Nice work @ujfalusi

File: sound/soc/sof/intel/hda-dsp.c [hunk 1]
```diff
@@ -1437,12 +1490,19 @@ void hda_dsp_get_state(struct snd_sof_dev *sdev, const char *level)
...
-	if (module == FSR_MOD_BRNGUP)
+	if (module == FSR_MOD_BRNGUP) {
 		state_text = hda_dsp_get_state_text(state, fsr_bringup_state_names,
 						    ARRAY_SIZE(fsr_bringup_state_names));
-	else
-		state_text = hda_dsp_get_state_text(state, fsr_rom_state_names,
-						    ARRAY_SIZE(fsr_rom_state_names));
+	} else {
+		if (chip->hw_ip_version < SOF_INTEL_ACE_1_0)
+			state_text = hda_dsp_get_state_text(state,
```
Comment (RanderWang) [inline sound/soc/sof/intel/hda-dsp.c:1498]: I have an idea that we build a array for each ip version and then we use  fsr_rom_state_names[chip->hw_ip_version].  Then we only need to add table and not change code logic

**Review by RanderWang — APPROVED** [review]
LGTM

File: sound/soc/sof/intel/hda-dsp.c (hunk 1 above)
Comment (ujfalusi) [reply sound/soc/sof/intel/hda-dsp.c:1498]: so far we have two version, one for <= CAVS_2_5 and one for >= ACE_1_0.
At this point I would not bother with that, but as soon as we need another state table we should do that, I agree.

**Comment by plbossart** [comment]
@ujfalusi can you recheck the minor checkpatch warnings? It's a really good PR, let's get rid of typos to make it shiny clean.

**Comment by ujfalusi** [comment]
Changes since v6:
- typos fixed in commit messages
- alignment checkpatch issues remained to make the code readable

**Review by lgirdwood — APPROVED** [review]
@plbossart good to merge ?

**Pull request merged**